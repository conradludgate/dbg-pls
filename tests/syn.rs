macro_rules! assert_pretty_syn_snapshot {
    ($ty:ty => $code:literal) => {
        let parsed = &syn::parse_str::<$ty>($code).unwrap();
        let format = format!(
            "pretty:\n{}\n\ncolor:\n{}\n",
            dbg_pls::pretty(&parsed),
            dbg_pls::color(&parsed)
        );
        insta::assert_snapshot!(insta::_macro_support::AutoName, &format, $code);
    };
}

#[test]
fn expr_array() {
    assert_pretty_syn_snapshot!(syn::Expr => r#"
            [
                "Hello, World! I am a long string",
                420,
                "Wait, you can't mix and match types in arrays, is this python?",
                69,
                "Nice."
            ]
        "#);
}

#[test]
fn qself() {
    assert_pretty_syn_snapshot!(syn::Expr => "<Foo as Bar::<<Baz as Qux>::Bang>>::Boo");
}

#[test]
fn item_macro() {
    assert_pretty_syn_snapshot!(syn::Item => r#"
            macro_rules! assert_pretty_syn_snapshot {
                ($ty:ty => $code:literal) => {
                    insta::assert_snapshot!(
                        insta::_macro_support::AutoName,
                        &crate::pretty(&syn::parse_str::<$ty>($code).unwrap()).to_string(),
                        $code
                    );
                };
            }
        "#);
}
