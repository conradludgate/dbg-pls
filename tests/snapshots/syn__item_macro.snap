---
source: tests/syn.rs
expression: "\n            macro_rules! assert_pretty_syn_snapshot {\n                ($ty:ty => $code:literal) => {\n                    insta::assert_snapshot!(\n                        insta::_macro_support::AutoName,\n                        &crate::pretty(&syn::parse_str::<$ty>($code).unwrap()).to_string(),\n                        $code\n                    );\n                };\n            }\n        "
---
pretty:
Macro(ItemMacro {
    attrs: [],
    ident: Some(assert_pretty_syn_snapshot),
    mac: Macro {
        path: Path {
            leading_colon: None,
            segments: [
                PathSegment {
                    ident: macro_rules,
                    arguments: None,
                },
            ],
        },
        bang_token: Not,
        delimiter: Brace(Brace),
        tokens: [
            Group(Group {
                delimiter: Parenthesis,
                stream: [
                    Punct(Punct { ch: '$', spacing: Alone }),
                    Ident(ty),
                    Punct(Punct { ch: ':', spacing: Alone }),
                    Ident(ty),
                    Punct(Punct { ch: '=', spacing: Joint }),
                    Punct(Punct { ch: '>', spacing: Alone }),
                    Punct(Punct { ch: '$', spacing: Alone }),
                    Ident(code),
                    Punct(Punct { ch: ':', spacing: Alone }),
                    Ident(literal),
                ],
            }),
            Punct(Punct { ch: '=', spacing: Joint }),
            Punct(Punct { ch: '>', spacing: Alone }),
            Group(Group {
                delimiter: Brace,
                stream: [
                    Ident(insta),
                    Punct(Punct { ch: ':', spacing: Joint }),
                    Punct(Punct { ch: ':', spacing: Alone }),
                    Ident(assert_snapshot),
                    Punct(Punct { ch: '!', spacing: Alone }),
                    Group(Group {
                        delimiter: Parenthesis,
                        stream: [
                            Ident(insta),
                            Punct(Punct { ch: ':', spacing: Joint }),
                            Punct(Punct { ch: ':', spacing: Alone }),
                            Ident(_macro_support),
                            Punct(Punct { ch: ':', spacing: Joint }),
                            Punct(Punct { ch: ':', spacing: Alone }),
                            Ident(AutoName),
                            Punct(Punct { ch: ',', spacing: Alone }),
                            Punct(Punct { ch: '&', spacing: Alone }),
                            Ident(crate),
                            Punct(Punct { ch: ':', spacing: Joint }),
                            Punct(Punct { ch: ':', spacing: Alone }),
                            Ident(pretty),
                            Group(Group {
                                delimiter: Parenthesis,
                                stream: [
                                    Punct(Punct { ch: '&', spacing: Alone }),
                                    Ident(syn),
                                    Punct(Punct { ch: ':', spacing: Joint }),
                                    Punct(Punct { ch: ':', spacing: Alone }),
                                    Ident(parse_str),
                                    Punct(Punct { ch: ':', spacing: Joint }),
                                    Punct(Punct { ch: ':', spacing: Joint }),
                                    Punct(Punct { ch: '<', spacing: Joint }),
                                    Punct(Punct { ch: '$', spacing: Alone }),
                                    Ident(ty),
                                    Punct(Punct { ch: '>', spacing: Alone }),
                                    Group(Group {
                                        delimiter: Parenthesis,
                                        stream: [
                                            Punct(Punct { ch: '$', spacing: Alone }),
                                            Ident(code),
                                        ],
                                    }),
                                    Punct(Punct { ch: '.', spacing: Alone }),
                                    Ident(unwrap),
                                    Group(Group {
                                        delimiter: Parenthesis,
                                        stream: [],
                                    }),
                                ],
                            }),
                            Punct(Punct { ch: '.', spacing: Alone }),
                            Ident(to_string),
                            Group(Group {
                                delimiter: Parenthesis,
                                stream: [],
                            }),
                            Punct(Punct { ch: ',', spacing: Alone }),
                            Punct(Punct { ch: '$', spacing: Alone }),
                            Ident(code),
                        ],
                    }),
                    Punct(Punct { ch: ';', spacing: Alone }),
                ],
            }),
            Punct(Punct { ch: ';', spacing: Alone }),
        ],
    },
    semi_token: None,
})

color:
[32mMacro[0m[31m([0m[32mItemMacro[0m [31m{[0m
    attrs[31m:[0m [31m[[0m[31m][0m[31m,[0m
    ident[31m:[0m Some[31m([0massert_pretty_syn_snapshot[31m)[0m[31m,[0m
    mac[31m:[0m [32mMacro[0m [31m{[0m
        path[31m:[0m [32mPath[0m [31m{[0m
            leading_colon[31m:[0m None[31m,[0m
            segments[31m:[0m [31m[[0m
                [32mPathSegment[0m [31m{[0m
                    ident[31m:[0m macro_rules[31m,[0m
                    arguments[31m:[0m None[31m,[0m
                [31m}[0m[31m,[0m
            [31m][0m[31m,[0m
        [31m}[0m[31m,[0m
        bang_token[31m:[0m [32mNot[0m[31m,[0m
        delimiter[31m:[0m [32mBrace[0m[31m([0m[32mBrace[0m[31m)[0m[31m,[0m
        tokens[31m:[0m [31m[[0m
            [32mGroup[0m[31m([0m[32mGroup[0m [31m{[0m
                delimiter[31m:[0m [32mParenthesis[0m[31m,[0m
                stream[31m:[0m [31m[[0m
                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m$[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                    [32mIdent[0m[31m([0mty[31m)[0m[31m,[0m
                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m:[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                    [32mIdent[0m[31m([0mty[31m)[0m[31m,[0m
                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m=[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mJoint[0m [31m}[0m[31m)[0m[31m,[0m
                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m>[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m$[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                    [32mIdent[0m[31m([0mcode[31m)[0m[31m,[0m
                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m:[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                    [32mIdent[0m[31m([0mliteral[31m)[0m[31m,[0m
                [31m][0m[31m,[0m
            [31m}[0m[31m)[0m[31m,[0m
            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m=[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mJoint[0m [31m}[0m[31m)[0m[31m,[0m
            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m>[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
            [32mGroup[0m[31m([0m[32mGroup[0m [31m{[0m
                delimiter[31m:[0m [32mBrace[0m[31m,[0m
                stream[31m:[0m [31m[[0m
                    [32mIdent[0m[31m([0minsta[31m)[0m[31m,[0m
                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m:[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mJoint[0m [31m}[0m[31m)[0m[31m,[0m
                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m:[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                    [32mIdent[0m[31m([0massert_snapshot[31m)[0m[31m,[0m
                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m![0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                    [32mGroup[0m[31m([0m[32mGroup[0m [31m{[0m
                        delimiter[31m:[0m [32mParenthesis[0m[31m,[0m
                        stream[31m:[0m [31m[[0m
                            [32mIdent[0m[31m([0minsta[31m)[0m[31m,[0m
                            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m:[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mJoint[0m [31m}[0m[31m)[0m[31m,[0m
                            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m:[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                            [32mIdent[0m[31m([0m_macro_support[31m)[0m[31m,[0m
                            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m:[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mJoint[0m [31m}[0m[31m)[0m[31m,[0m
                            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m:[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                            [32mIdent[0m[31m([0m[32mAutoName[0m[31m)[0m[31m,[0m
                            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m,[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m&[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                            [32mIdent[0m[31m([0m[31mcrate[0m[31m)[0m[31m,[0m
                            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m:[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mJoint[0m [31m}[0m[31m)[0m[31m,[0m
                            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m:[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                            [32mIdent[0m[31m([0mpretty[31m)[0m[31m,[0m
                            [32mGroup[0m[31m([0m[32mGroup[0m [31m{[0m
                                delimiter[31m:[0m [32mParenthesis[0m[31m,[0m
                                stream[31m:[0m [31m[[0m
                                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m&[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                                    [32mIdent[0m[31m([0msyn[31m)[0m[31m,[0m
                                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m:[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mJoint[0m [31m}[0m[31m)[0m[31m,[0m
                                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m:[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                                    [32mIdent[0m[31m([0mparse_str[31m)[0m[31m,[0m
                                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m:[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mJoint[0m [31m}[0m[31m)[0m[31m,[0m
                                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m:[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mJoint[0m [31m}[0m[31m)[0m[31m,[0m
                                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m<[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mJoint[0m [31m}[0m[31m)[0m[31m,[0m
                                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m$[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                                    [32mIdent[0m[31m([0mty[31m)[0m[31m,[0m
                                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m>[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                                    [32mGroup[0m[31m([0m[32mGroup[0m [31m{[0m
                                        delimiter[31m:[0m [32mParenthesis[0m[31m,[0m
                                        stream[31m:[0m [31m[[0m
                                            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m$[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                                            [32mIdent[0m[31m([0mcode[31m)[0m[31m,[0m
                                        [31m][0m[31m,[0m
                                    [31m}[0m[31m)[0m[31m,[0m
                                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m.[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                                    [32mIdent[0m[31m([0munwrap[31m)[0m[31m,[0m
                                    [32mGroup[0m[31m([0m[32mGroup[0m [31m{[0m
                                        delimiter[31m:[0m [32mParenthesis[0m[31m,[0m
                                        stream[31m:[0m [31m[[0m[31m][0m[31m,[0m
                                    [31m}[0m[31m)[0m[31m,[0m
                                [31m][0m[31m,[0m
                            [31m}[0m[31m)[0m[31m,[0m
                            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m.[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                            [32mIdent[0m[31m([0mto_string[31m)[0m[31m,[0m
                            [32mGroup[0m[31m([0m[32mGroup[0m [31m{[0m
                                delimiter[31m:[0m [32mParenthesis[0m[31m,[0m
                                stream[31m:[0m [31m[[0m[31m][0m[31m,[0m
                            [31m}[0m[31m)[0m[31m,[0m
                            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m,[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m$[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                            [32mIdent[0m[31m([0mcode[31m)[0m[31m,[0m
                        [31m][0m[31m,[0m
                    [31m}[0m[31m)[0m[31m,[0m
                    [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m;[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
                [31m][0m[31m,[0m
            [31m}[0m[31m)[0m[31m,[0m
            [32mPunct[0m[31m([0m[32mPunct[0m [31m{[0m ch[31m:[0m [33m[31m'[33m;[0m[33m[31m'[0m[31m,[0m spacing[31m:[0m [32mAlone[0m [31m}[0m[31m)[0m[31m,[0m
        [31m][0m[31m,[0m
    [31m}[0m[31m,[0m
    semi_token[31m:[0m None[31m,[0m
[31m}[0m[31m)[0m


